File: src/data/tutorials/006-maximum-subarray.js
/**
 * Tutorial content for #006 — Maximum Subarray
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an integer array (which may contain negative numbers), find the contiguous subarray that has the largest sum and return that sum. The subarray must contain at least one element.`,
    example: `nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]\n→ The subarray [4, -1, 2, 1] has sum = 6\n✅ Answer: 6`,
    keyInsight: `If the running sum ever drops below zero, it's dead weight — any future subarray is better off starting fresh. This is the heart of Kadane's Algorithm.`,
  },

  approaches: {
    'Brute Force': {
      intuition: `Try every possible starting index and every possible ending index. For each pair, sum all elements between them and track the maximum. Three nested loops: two to define start/end, one to compute the sum.`,
      steps: [
        `Initialize maxSum = Integer.MIN_VALUE.`,
        `Outer loop i from 0 to n-1 (start of subarray).`,
        `Inner loop j from i to n-1 (end of subarray).`,
        `Innermost loop k from i to j: sum up nums[i..j].`,
        `Update maxSum = max(maxSum, current sum).`,
        `Return maxSum.`,
      ],
      example: `nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]\n\nstart=0: [-2]=−2, [-2,1]=−1, [-2,1,−3]=−4, ...\nstart=3: [4]=4, [4,−1]=3, [4,−1,2]=5, [4,−1,2,1]=6 ← max!\n...\n✅ Answer: 6`,
      keyInsight: `O(n³) time, O(1) space. Exhaustive but very slow. Only practical for n < 200.`,
    },

    'Optimized Brute Force': {
      intuition: `Instead of a third inner loop to recompute the sum, extend the running sum as you expand the right boundary. This cuts one loop without changing the two-pointer idea: fix the start, try all ends and accumulate.`,
      steps: [
        `Initialize maxSum = Integer.MIN_VALUE.`,
        `Outer loop i from 0 to n-1 (start index).`,
        `Initialize currentSum = 0.`,
        `Inner loop j from i to n-1: currentSum += nums[j].`,
        `Update maxSum = max(maxSum, currentSum).`,
        `Return maxSum.`,
      ],
      example: `nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]\n\ni=3 (start=4):\n  j=3: sum=4,  max=4\n  j=4: sum=3,  max=4\n  j=5: sum=5,  max=5\n  j=6: sum=6,  max=6 ← best\n  j=7: sum=1,  max=6\n  j=8: sum=5,  max=6\n✅ Answer: 6`,
      keyInsight: `O(n²) time, O(1) space. Twice as fast as O(n³) in practice, but still quadratic. Good stepping stone toward Kadane's.`,
    },

    "Kadane's Algorithm": {
      intuition: `Walk through the array once, maintaining a "current subarray sum". At each element, decide: is it better to extend the existing subarray (add this element) or start a brand new subarray here? If the current sum is negative, starting fresh is always better. Track the best sum seen at any point.`,
      steps: [
        `Initialize currentSum = nums[0], maxSum = nums[0].`,
        `Loop i from 1 to n-1.`,
        `currentSum = max(nums[i], currentSum + nums[i]).`,
        `  (This is "start fresh if negative, else extend")`,
        `maxSum = max(maxSum, currentSum).`,
        `Return maxSum.`,
      ],
      example: `nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]\n\ni=0: cur=-2,  max=-2\ni=1: cur=max(1,-2+1)=1,   max=1\ni=2: cur=max(-3,1-3)=-2,  max=1\ni=3: cur=max(4,-2+4)=4,   max=4\ni=4: cur=max(-1,4-1)=3,   max=4\ni=5: cur=max(2,3+2)=5,    max=5\ni=6: cur=max(1,5+1)=6,    max=6\ni=7: cur=max(-5,6-5)=1,   max=6\ni=8: cur=max(4,1+4)=5,    max=6\n✅ Answer: 6`,
      keyInsight: `O(n) time, O(1) space — optimal. One of the most elegant algorithms in computer science. The decision at each step (extend or restart) is greedy and provably correct.`,
    },

    "Kadane's with Subarray Indices": {
      intuition: `Same as Kadane's, but also track where the best subarray starts and ends. Whenever we restart (current sum was negative), record the new start index. Whenever we update the global max, record the current start and end indices.`,
      steps: [
        `Initialize currentSum = nums[0], maxSum = nums[0].`,
        `Track start=0, end=0, tempStart=0.`,
        `Loop i from 1 to n-1:`,
        `  If currentSum < 0: currentSum = nums[i], tempStart = i (restart here).`,
        `  Else: currentSum += nums[i].`,
        `  If currentSum > maxSum: maxSum = currentSum, start = tempStart, end = i.`,
        `Return maxSum (and optionally nums[start..end]).`,
      ],
      example: `nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]\n\ni=1: cur<0→restart, cur=1, tempStart=1, maxSum=1, start=1,end=1\ni=2: cur=1-3=-2, maxSum still 1\ni=3: cur<0→restart, cur=4, tempStart=3, maxSum=4, start=3,end=3\ni=4: cur=3, maxSum=4\ni=5: cur=5, maxSum=5, start=3,end=5\ni=6: cur=6, maxSum=6, start=3,end=6\n...\n✅ Subarray: nums[3..6] = [4,-1,2,1], sum=6`,
      keyInsight: `Same O(n) time, O(1) space as Kadane's but records the actual subarray. Useful when you need to return the subarray itself, not just the sum.`,
    },

    'Divide and Conquer': {
      intuition: `Split the array in half recursively. The maximum subarray either lives entirely in the left half, entirely in the right half, or crosses the midpoint. The crossing case is handled by expanding outward from the midpoint in both directions and summing the best halves. Combine the three candidates and return the max.`,
      steps: [
        `Base case: if left == right, return nums[left].`,
        `Compute mid = left + (right - left) / 2.`,
        `Recursively find maxLeft = solve(nums, left, mid) and maxRight = solve(nums, mid+1, right).`,
        `Compute crossMax: expand left from mid and right from mid+1, accumulating sums, track best each direction.`,
        `Return max(maxLeft, maxRight, crossMax).`,
      ],
      example: `nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]\nmid = 4 (value -1)\n\nLeft half [-2,1,-3,4,-1]: recursion → 4\nRight half [2,1,-5,4]: recursion → 3\nCross: expand left from mid=4: -1,(-1+4)=3,(3-3)=0... best=3\n       expand right from mid+1=5: 2,(2+1)=3... best=3\n       crossMax = 3+3 = 6 ← wins!\n✅ Answer: 6`,
      keyInsight: `O(n log n) time, O(log n) stack space. Elegant divide-and-conquer but strictly worse than Kadane's O(n). Valuable academically to demonstrate the paradigm.`,
    },
  },
}

File: public/cpp/006-maximum-subarray.cpp
/*
 * LeetCode Problem #53: Maximum Subarray (Kadane's Algorithm)
 * Difficulty: Easy
 * Link: https://leetcode.com/problems/maximum-subarray/
 * 
 * Problem Statement:
 * Given an integer array nums, find the contiguous subarray (containing at least
 * one number) which has the largest sum and return its sum.
 */

#include <vector>
#include <algorithm>
#include <climits>
#include <iostream>
using namespace std;

class Solution {
public:
    // ==================== APPROACH 1: Brute Force ====================
    /*
     * Algorithm:
     * - Check all possible subarrays
     * - Calculate sum for each subarray
     * - Track maximum sum
     * 
     * Time Complexity: O(n³) - three nested loops
     * Space Complexity: O(1)
     * 
     * When to use: Never (too slow)
     */
    int maxSubArray_BruteForce(vector<int>& nums) {
        int n = nums.size();
        int maxSum = INT_MIN;
        
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int sum = 0;
                for (int k = i; k <= j; k++) {
                    sum += nums[k];
                }
                maxSum = max(maxSum, sum);
            }
        }
        
        return maxSum;
    }
    
    // ==================== APPROACH 2: Optimized Brute Force ====================
    /*
     * Algorithm:
     * - Calculate sum incrementally instead of recalculating
     * - For each starting position, extend subarray and track sum
     * 
     * Time Complexity: O(n²)
     * Space Complexity: O(1)
     * 
     * When to use: Better than O(n³) but still not optimal
     */
    int maxSubArray_OptimizedBrute(vector<int>& nums) {
        int n = nums.size();
        int maxSum = INT_MIN;
        
        for (int i = 0; i < n; i++) {
            int currentSum = 0;
            for (int j = i; j < n; j++) {
                currentSum += nums[j];
                maxSum = max(maxSum, currentSum);
            }
        }
        
        return maxSum;
    }
    
    // ==================== APPROACH 3: Kadane's Algorithm (Optimal) ====================
    /*
     * Algorithm:
     * - Track current subarray sum
     * - If current sum becomes negative, start new subarray
     * - Keep track of maximum sum seen so far
     * 
     * Time Complexity: O(n) - single pass
     * Space Complexity: O(1)
     * 
     * When to use: This is the OPTIMAL solution
     * 
     * Key Insight:
     * - If current sum is negative, it won't help future elements
     * - Better to start fresh from next element
     * - At each position, decide: extend current subarray or start new one
     * 
     * Decision at each step:
     * - currentSum = max(nums[i], currentSum + nums[i])
     * - This means: either start new subarray at i, or extend previous
     */
    int maxSubArray_Kadane(vector<int>& nums) {
        int maxSum = nums[0];
        int currentSum = nums[0];
        
        for (int i = 1; i < nums.size(); i++) {
            // Either extend current subarray or start new one
            currentSum = max(nums[i], currentSum + nums[i]);
            maxSum = max(maxSum, currentSum);
        }
        
        return maxSum;
    }
    
    // ==================== APPROACH 4: Kadane's with Subarray Indices ====================
    /*
     * Same as Kadane's but also tracks start and end indices
     * 
     * Time Complexity: O(n)
     * Space Complexity: O(1)
     * 
     * When to use: When you need to return the actual subarray
     */
    int maxSubArray_KadaneWithIndices(vector<int>& nums, int& start, int& end) {
        int maxSum = nums[0];
        int currentSum = nums[0];
        start = 0;
        end = 0;
        int tempStart = 0;
        
        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] > currentSum + nums[i]) {
                currentSum = nums[i];
                tempStart = i;  // New subarray starts here
            } else {
                currentSum = currentSum + nums[i];
            }
            
            if (currentSum > maxSum) {
                maxSum = currentSum;
                start = tempStart;
                end = i;
            }
        }
        
        return maxSum;
    }
    
    // ==================== APPROACH 5: Divide and Conquer ====================
    /*
     * Algorithm:
     * - Divide array into two halves
     * - Maximum subarray is either:
     *   1. In left half
     *   2. In right half
     *   3. Crosses the middle
     * - Recursively solve and combine
     * 
     * Time Complexity: O(n log n)
     * Space Complexity: O(log n) - recursion stack
     * 
     * When to use: Academic interest, not optimal for this problem
     */
    int maxCrossingSum(vector<int>& nums, int left, int mid, int right) {
        int leftSum = INT_MIN;
        int sum = 0;
        for (int i = mid; i >= left; i--) {
            sum += nums[i];
            leftSum = max(leftSum, sum);
        }
        
        int rightSum = INT_MIN;
        sum = 0;
        for (int i = mid + 1; i <= right; i++) {
            sum += nums[i];
            rightSum = max(rightSum, sum);
        }
        
        return leftSum + rightSum;
    }
    
    int maxSubArrayHelper(vector<int>& nums, int left, int right) {
        if (left == right) return nums[left];
        
        int mid = left + (right - left) / 2;
        
        int leftMax = maxSubArrayHelper(nums, left, mid);
        int rightMax = maxSubArrayHelper(nums, mid + 1, right);
        int crossMax = maxCrossingSum(nums, left, mid, right);
        
        return max({leftMax, rightMax, crossMax});
    }
    
    int maxSubArray_DivideConquer(vector<int>& nums) {
        return maxSubArrayHelper(nums, 0, nums.size() - 1);
    }
};

// ==================== TEST CASES ====================
void runTests() {
    Solution sol;
    
    // Test Case 1: Standard case
    vector<int> nums1 = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    cout << "Test 1: " << sol.maxSubArray_Kadane(nums1) << endl;
    // Expected: 6 (subarray [4,-1,2,1])
    
    // Test Case 2: All negative
    vector<int> nums2 = {-2, -3, -1, -4};
    cout << "Test 2: " << sol.maxSubArray_Kadane(nums2) << endl;
    // Expected: -1 (single element)
    
    // Test Case 3: Single element
    vector<int> nums3 = {5};
    cout << "Test 3: " << sol.maxSubArray_Kadane(nums3) << endl;
    // Expected: 5
    
    // Test Case 4: All positive
    vector<int> nums4 = {1, 2, 3, 4, 5};
    cout << "Test 4: " << sol.maxSubArray_Kadane(nums4) << endl;
    // Expected: 15 (entire array)
    
    // Test Case 5: With indices
    vector<int> nums5 = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    int start, end;
    int maxSum = sol.maxSubArray_KadaneWithIndices(nums5, start, end);
    cout << "Test 5: Sum=" << maxSum << ", Start=" << start << ", End=" << end << endl;
    // Expected: Sum=6, Start=3, End=6
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPARISON OF APPROACHES:
 * 
 * 1. Brute Force:
 *    Time: O(n³), Space: O(1)
 *    Not acceptable
 * 
 * 2. Optimized Brute:
 *    Time: O(n²), Space: O(1)
 *    Better but still slow
 * 
 * 3. Kadane's Algorithm (RECOMMENDED):
 *    Time: O(n), Space: O(1)
 *    Optimal solution, most elegant
 * 
 * 4. Divide and Conquer:
 *    Time: O(n log n), Space: O(log n)
 *    Academic interest, not optimal
 * 
 * INTERVIEW TIPS:
 * - This is a classic DP problem
 * - Kadane's algorithm is the expected solution
 * - Explain the greedy choice: extend or start new
 * - Handle all negative numbers case
 * - Can extend to return actual subarray
 * 
 * KADANE'S ALGORITHM INTUITION:
 * - At each position, we have two choices:
 *   1. Add current element to existing subarray
 *   2. Start new subarray from current element
 * - We choose whichever gives larger sum
 * - If adding current element makes sum negative, start fresh
 * 
 * STEP-BY-STEP for [-2,1,-3,4,-1,2,1,-5,4]:
 * 
 * i=0: current=-2, max=-2
 * i=1: current=1 (start new), max=1
 * i=2: current=-2 (1-3), max=1
 * i=3: current=4 (start new), max=4
 * i=4: current=3 (4-1), max=4
 * i=5: current=5 (3+2), max=5
 * i=6: current=6 (5+1), max=6 ✓
 * i=7: current=1 (6-5), max=6
 * i=8: current=5 (1+4), max=6
 * 
 * COMMON MISTAKES:
 * - Not handling all negative numbers
 * - Resetting current sum to 0 instead of current element
 * - Not updating max at each step
 * - Confusing with maximum product subarray
 * 
 * FOLLOW-UP QUESTIONS:
 * - Return the actual subarray? (Track indices)
 * - Maximum product subarray? (Different algorithm - LeetCode #152)
 * - Circular array? (Two passes or break circle)
 * - 2D array? (Extend to 2D Kadane's)
 * 
 * VARIATIONS:
 * - Maximum Subarray Sum with One Deletion (LeetCode #1186)
 * - Maximum Sum Circular Subarray (LeetCode #918)
 * - Maximum Product Subarray (LeetCode #152)
 * 
 * WHY KADANE'S WORKS:
 * - Optimal substructure: optimal solution contains optimal subsolutions
 * - Greedy choice: at each step, make locally optimal choice
 * - If current sum is negative, it can't help future elements
 * - This greedy choice leads to global optimum
 */

// Made with Bob
